src/vs/workbench/services/semanticSimilarity/common/semanticSimilarityService.ts

```diff
@@ -5,25 +5,118 @@
 
 import { CommandsRegistry, ICommandService } from 'vs/platform/commands/common/commands';
 import { IFileService } from 'vs/platform/files/common/files';
-import { InstantiationType, registerSingleton } from 'vs/platform/instantiation/common/extensions';
 import { createDecorator } from 'vs/platform/instantiation/common/instantiation';
 import { URI } from 'vs/base/common/uri';
 import { CancellationToken } from 'vs/base/common/cancellation';
-import { raceCancellation } from 'vs/base/common/async';
+import { CancelablePromise, createCancelablePromise, raceCancellablePromises, raceCancellation, timeout } from 'vs/base/common/async';
+import { IDisposable } from 'vs/base/common/lifecycle';
+import { InstantiationType, registerSingleton } from 'vs/platform/instantiation/common/extensions';
 
-export const ISemanticSimilarityService = createDecorator<ISemanticSimilarityService>('IEmbeddingsService');
+export const ISemanticSimilarityService = createDecorator<ISemanticSimilarityService>('ISemanticSimilarityService');
 
 export interface ISemanticSimilarityService {
+	readonly _serviceBrand: undefined;
+
 	isEnabled(): boolean;
 	getSimilarityScore(string1: string, comparisons: string[], token: CancellationToken): Promise<number[]>;
+	registerSemanticSimilarityProvider(provider: ISemanticSimilarityProvider): IDisposable;
+	unregisterSemanticSimilarityProvider(provider: ISemanticSimilarityProvider): void;
+}
+
+export interface ISemanticSimilarityProvider {
+	provideSimilarityScore(string1: string, comparisons: string[], token: CancellationToken): Promise<number[]>;
 }
 
+export class SemanticSimilarityService implements ISemanticSimilarityService {
+	readonly _serviceBrand: undefined;
+
+	static readonly DEFAULT_TIMEOUT = 1000 * 10; // 30 seconds
+
+	private readonly _providers: ISemanticSimilarityProvider[] = [];
+	// remove when we move over to API
+	private readonly oldService: OldSemanticSimilarityService;
+
+	constructor(
+		// for the old service
+		@ICommandService commandService: ICommandService,
+		@IFileService fileService: IFileService
+	) {
+		this.oldService = new OldSemanticSimilarityService(commandService, fileService);
+	}
+
+	isEnabled(): boolean {
+		return this._providers.length > 0;
+	}
+
+	registerSemanticSimilarityProvider(provider: ISemanticSimilarityProvider): IDisposable {
+		this._providers.push(provider);
+		return {
+			dispose: () => {
+				const index = this._providers.indexOf(provider);
+				if (index >= 0) {
+					this._providers.splice(index, 1);
+				}
+			}
+		};
+	}
+
+	unregisterSemanticSimilarityProvider(provider: ISemanticSimilarityProvider): void {
```

Comment:
You have this and also a disposable returned from the register method, I think either is fine